### PR TITLE
feat: dao based governance NTRN-206

### PR DESCRIPTION
**Pull request by quasisamurai**
[Task 1](https://p2pvalidator.atlassian.net/browse/NTRN-169)
[Task 2](https://p2pvalidator.atlassian.net/browse/NTRN-206)

This PR implements a neutron's vision of governance. It uses [admin module](https://github.com/cosmos/composer) in bound with smart contract bindings, provided w this pr.

testing: 
- `make init`
- from [this pr](https://github.com/neutron-org/neutron-dao/pull/8) `./test_proposal.sh`

### CODE REVIEW

File: proto/adminmodule/genesis.proto [hunk 1]
```diff
@@ -0,0 +1,13 @@
+syntax = "proto3";
+package cosmos.adminmodule.adminmodule;
+
+import "gogoproto/gogo.proto";
+
+// this line is used by starport scaffolding # genesis/proto/import
```
Comment (zavgorodnii) [inline proto/adminmodule/genesis.proto]: pls remove

File: proto/adminmodule/query.proto [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+import "google/api/annotations.proto";
+import "cosmos/gov/v1beta1/gov.proto";
+// this line is used by starport scaffolding # 1
+
+option go_package = "github.com/cosmos/admin-module/x/adminmodule/types";
+
+// Query defines the gRPC querier service.
+service Query {
+  // this line is used by starport scaffolding # 2
```
Comment (zavgorodnii) [inline proto/adminmodule/query.proto]: all these likes should be removed i think

File: wasmbinding/bindings/msg.go [hunk 3]
```diff
@@ -17,6 +19,8 @@ type NeutronMsg struct {
 	RegisterInterchainQuery   *RegisterInterchainQuery   `json:"register_interchain_query,omitempty"`
 	UpdateInterchainQuery     *UpdateInterchainQuery     `json:"update_interchain_query,omitempty"`
 	RemoveInterchainQuery     *RemoveInterchainQuery     `json:"remove_interchain_query,omitempty"`
+	AddAdmin                  *AddAdmin                  `json:"add_admin,omitempty"`
```
Comment (zavgorodnii) [inline wasmbinding/bindings/msg.go]: why exactly do we need an add admin message available for the contracts? aren't we going to add a single admin to genesis?

File: wasmbinding/message_plugin.go [hunk 4]
```diff
@@ -193,6 +206,138 @@ func (m *CustomMessenger) submitTx(ctx sdk.Context, contractAddr sdk.AccAddress,
...
+	if submitProposal.Proposals.TextProposal != nil {
+
+		prop := govtypes.TextProposal{
+			Title:       submitProposal.Proposals.TextProposal.Title,
+			Description: submitProposal.Proposals.TextProposal.Description,
+		}
+		cont, err := proto.Marshal(&prop)
+		if err != nil {
+			return nil, sdkerrors.Wrap(err, "failed to marshall incoming SubmitProposal message")
+		}
```
Comment (zavgorodnii) [inline wasmbinding/message_plugin.go:245]: here and everywhere: please add newlines after error checks

File: wasmbinding/message_plugin.go [hunk 5]
```diff
@@ -193,6 +206,138 @@ func (m *CustomMessenger) submitTx(ctx sdk.Context, contractAddr sdk.AccAddress,
...
+		"creator", contractAddr.String(),
+	)
+	return nil, [][]byte{data}, nil
+}
+
+func (m *CustomMessenger) PerformSubmitProposal(ctx sdk.Context, contractAddr sdk.AccAddress, submitProposal *bindings.SubmitProposal) (*admintypes.MsgSubmitProposalResponse, error) {
+	msg := admintypes.MsgSubmitProposal{Proposer: contractAddr.String()}
+
+	if submitProposal.Proposals.TextProposal != nil {
+
```
Comment (zavgorodnii) [inline wasmbinding/message_plugin.go]: redundant newline

File: wasmbinding/message_plugin.go [hunk 6]
```diff
@@ -193,6 +206,138 @@ func (m *CustomMessenger) submitTx(ctx sdk.Context, contractAddr sdk.AccAddress,
...
+		return nil, sdkerrors.Wrap(err, "failed to submit proposal")
+	}
+
+	return response, nil
+}
+
+func (m *CustomMessenger) addAmin(ctx sdk.Context, contractAddr sdk.AccAddress, addAdmin *bindings.AddAdmin) ([]sdk.Event, [][]byte, error) {
+	response, err := m.PerformAddAmin(ctx, contractAddr, addAdmin)
+	if err != nil {
+		ctx.Logger().Debug("PerformSubmitTx: failed to addAdmin",
```
Comment (zavgorodnii) [inline wasmbinding/message_plugin.go]: `PerformSubmitTx` is a copy-and-paste

File: wasmbinding/wasm.go [hunk 7]
```diff
@@ -3,20 +3,21 @@ package wasmbinding
...
-func RegisterCustomPlugins(ictxKeeper *interchaintransactionsmodulekeeper.Keeper, icqKeeper *interchainqueriesmodulekeeper.Keeper) []wasmkeeper.Option {
+func RegisterCustomPlugins(ictxKeeper *interchaintransactionsmodulekeeper.Keeper, icqKeeper *interchainqueriesmodulekeeper.Keeper, admKeeper *adminmodulemodulekeeper.Keeper) []wasmkeeper.Option {
 	wasmQueryPlugin := NewQueryPlugin(ictxKeeper, icqKeeper)
 
 	queryPluginOpt := wasmkeeper.WithQueryPlugins(&wasmkeeper.QueryPlugins{
 		Custom: CustomQuerier(wasmQueryPlugin),
 	})
 	messagePluginOpt := wasmkeeper.WithMessageHandlerDecorator(
-		CustomMessageDecorator(ictxKeeper, icqKeeper),
+		CustomMessageDecorator(ictxKeeper, icqKeeper, admKeeper),
```
Comment (zavgorodnii) [inline wasmbinding/wasm.go:20]: maybe give this variable a full name? `adminKeeper`? not sure we gain anything from losing a couple of characters. also in other places it has a full name

File: wasmbinding/bindings/msg.go (hunk 3 above)
Comment (quasisamurai) [reply wasmbinding/bindings/msg.go]: That was a very first method i implemented, so i have not removed `AddAdmin` exactly to leave a space to discussion - "are we need this one?".

File: x/adminmodule/abci.go [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+		keeper.AddToArchive(ctx, proposal)
+
+		logger.Info(
+			"proposal tallied",
+			"proposal", proposal.ProposalId,
+			"title", proposal.GetTitle(),
+			"result", logMsg,
+		)
+
+		// TODO event?
```
Comment (zavgorodnii) [inline x/adminmodule/abci.go:53]: add the event?

File: x/adminmodule/client/cli/query.go [hunk 9]
```diff
@@ -0,0 +1,32 @@
+package cli
+
+import (
+	"fmt"
+	// "strings"
+
+	"github.com/spf13/cobra"
+
+	"github.com/cosmos/cosmos-sdk/client"
+	// "github.com/cosmos/cosmos-sdk/client/flags"
```
Comment (zavgorodnii) [inline x/adminmodule/client/cli/query.go]: commented code

File: x/adminmodule/types/message_submit_proposal.go [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+	}
+
+	err := m.SetContent(content)
+	if err != nil {
+		return nil, err
+	}
+	return m, nil
+}
+
+//func (m *MsgSubmitProposal) GetContent() govtypes.Content { // TODO m.Content.GetCachedValue() returns nil!
```
Comment (zavgorodnii) [inline x/adminmodule/types/message_submit_proposal.go]: commented code

File: x/adminmodule/module.go [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+// AppModule implements the AppModule interface for the capability module.
+type AppModule struct {
+	AppModuleBasic
+
+	keeper keeper.Keeper
+	cdc    codec.JSONCodec
+}
+
+func (am AppModule) DefaultGenesis(jsonCodec codec.JSONCodec) json.RawMessage {
+	//TODO implement me
```
Comment (zavgorodnii) [inline x/adminmodule/module.go]: implement?

File: x/adminmodule/module.go [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+	panic("implement me")
+}
+
+func (am AppModule) ValidateGenesis(jsonCodec codec.JSONCodec, config client.TxEncodingConfig, message json.RawMessage) error {
+	//TODO implement me
```
Comment (zavgorodnii) [inline x/adminmodule/module.go]: implement?

File: x/adminmodule/keeper/querier.go [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+	abci "github.com/tendermint/tendermint/abci/types"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/neutron-org/neutron/x/adminmodule/types"
+)
+
+// NewQuerier creates a new adminmodule Querier instance
+func NewQuerier(keeper Keeper, legacyQuerierCdc *codec.LegacyAmino) sdk.Querier {
```
Comment (zavgorodnii) [inline x/adminmodule/keeper/querier.go:13]: i don't think this is the way to go in sdk 0.45; please check the query implementations in our modules

File: wasmbinding/message_plugin.go [hunk 14]
```diff
@@ -193,6 +206,138 @@ func (m *CustomMessenger) submitTx(ctx sdk.Context, contractAddr sdk.AccAddress,
...
+			Value:   cont,
+		}
+
+	} else if submitProposal.Proposals.ParamChangeProposal != nil {
+		proposal := submitProposal.Proposals.ParamChangeProposal
+		prop := paramChange.ParameterChangeProposal{
+			Title:       submitProposal.Proposals.TextProposal.Title,
+			Description: submitProposal.Proposals.TextProposal.Description,
+		}
+		var parameterChanges []paramChange.ParamChange
```
Comment (mmikev) [inline wasmbinding/message_plugin.go]: `parameterChanges := make([]paramChange.ParamChange, 0, len(proposal.Changes))`

File: wasmbinding/message_plugin.go [hunk 15]
```diff
@@ -193,6 +206,138 @@ func (m *CustomMessenger) submitTx(ctx sdk.Context, contractAddr sdk.AccAddress,
...
+		for _, parameterChange := range proposal.Changes {
```
Comment (mmikev) [inline wasmbinding/message_plugin.go]: What about making `proposal.Changes` in `bindings.SubmitProposal` as `paramChange.ParamChange` type?
You don't need this loop then.

**Review by pr0n00gler — CHANGES_REQUESTED** [review]
Please add tests for the new msg bindings

**Review by foxpy — CHANGES_REQUESTED** [review]
This branch breaks `interchain_tx_query` integration test: 
image

**Comment by zavgorodnii** [comment]
NOTE FOR EVERYONE: WE DO NOT NEED TO FIX MODULE FORMATTING, ETC.; WE'LL USE [INFORMAL](https://github.com/cosmos/interchain-security/tree/ethernal/provider-governance/x/ccv/adminmodule) ADMIN MODULE LATER ON.

**Review by foxpy — CHANGES_REQUESTED** [review]
Please rebase on upstream audit fixes branch since you lag a lot behind it.

**Review by pr0n00gler — DISMISSED** [review]
utACK

**Pull request closed without merging**